app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx

```diff
@@ -1,48 +1,232 @@
 import * as React from 'react';
-import { map } from 'lodash';
+import { countBy } from 'lodash';
+import { FieldArray } from 'formik';
 
-import { IStageConfigProps, StageConfigField } from 'core/pipeline';
-import { MapEditor } from 'core/forms';
+import { IStage } from 'core/domain';
+import { FormikStageConfig, IFormikStageConfigInjectedProps, IStageConfigProps } from 'core/pipeline';
+import {
+  FormValidator,
+  FormikFormField,
+  ILayoutProps,
+  IStageForSpelPreview,
+  IValidator,
+  LayoutContext,
+  LayoutProvider,
+  Markdown,
+  SpelInput,
+  StandardFieldLayout,
+  TextInput,
+  Tooltip,
+  ValidationMessage,
+  errorMessage,
+  useIsMountedRef,
+} from 'core/presentation';
+
+import { ExecutionAndStagePicker, IExecutionAndStagePickerProps } from './ExecutionAndStagePicker';
+
+import './EvaluateVariablesStageConfig.less';
 
 export interface IEvaluatedVariable {
   key: string;
   value: string;
 }
 
-export class EvaluateVariablesStageConfig extends React.Component<IStageConfigProps> {
-  private expand(variables: { [key: string]: string }): IEvaluatedVariable[] {
-    return map(variables, (value, key): IEvaluatedVariable => ({ key, value }));
-  }
-
-  private mapChanged = (key: string, values: { [key: string]: string }) => {
-    this.props.updateStageField({ [key]: this.expand(values) });
-  };
-
-  public render() {
-    const {
-      stage: { variables = [] },
-      pipeline,
-    } = this.props;
-
-    // Flattens an array of objects {key, value} into a single object with the respective keys/values
-    const variablesObject = variables.reduce(
-      // tslint:disable-next-line:prefer-object-spread
-      (acc: any, { key, value }: IEvaluatedVariable) => Object.assign(acc, { [key]: value }),
-      {},
-    );
-
-    return (
-      <div className="form-horizontal">
-        <StageConfigField label="Variables to evaluate">
-          <MapEditor
-            model={variablesObject}
-            allowEmpty={true}
-            onChange={(v: any) => this.mapChanged('variables', v)}
-            pipeline={pipeline}
-            valueCanContainSpel={true}
+const variableNameValidator: IValidator = (val: string, label: string) =>
```

Comment:
If there are other use cases for alphanumeric, snake case string validation then we should move this to the general validator file.

Reply:
+1 honestly even if there aren't use cases elsewhere we can think of off the top of our heads right now, I'd like to see this in the core set of validators as it's a pretty generic requirement (might even be applicable to things like app/stack/detail on infra?)